pliers/extractors/models.py

```diff
@@ -1,14 +1,242 @@
+                         listify)
+
+import logging
 
 tf = attempt_to_import('tensorflow')
+hub = attempt_to_import('tensorflow_hub')
 attempt_to_import('tensorflow.keras')
+attempt_to_import('tensorflow_text')
+
+
+class TFHubExtractor(Extractor):
+
+    ''' A generic class for Tensorflow Hub extractors 
+    Args:
+        url_or_path (str): url or path to TFHub model. You can
+            browse models at https://tfhub.dev/.
+        features (optional): list of labels (for classification) 
+            or other feature names. The number of items must 
+            match the number of features in the output. For example,
+            if a classification model with 1000 output classes is passed 
+            (e.g. EfficientNet B6, 
+            see https://tfhub.dev/tensorflow/efficientnet/b6/classification/1), 
+            this must be a list containing 1000 items. If a text encoder 
+            outputting 768-dimensional encoding is passed (e.g. base BERT),
+            this must be a list containing 768 items. Each dimension in the 
+            model output will be returned as a separate feature in the 
+            ExtractorResult.
+            Alternatively, the model output can be packed into a single 
+            feature (i.e. a vector) by passing a single-element list 
+            (e.g. ['encoding']) or a string. Along the lines of 
+            the previous examples, if a single feature name is 
+            passed here (e.g. if features=['encoding']) for a TFHub model 
+            that outputs a 768-dimensional encoding, the extractor will 
+            return only one feature named 'encoding', which contains the 
+            encoding vector as a 1-d array wrapped in a list.
+            If no value is passed, the extractor will automatically 
+            compute the number of features in the model output 
+            and return an equal number of features in pliers, labeling
+            each feature with a generic prefix + its positional index 
+            in the model output (feature_0, feature_1, ... ,feature_n).
+        transform_out (optional): function to transform model 
+            output for compatibility with extractor result
+        transform_inp (optional): function to transform Stim.data 
+            for compatibility with model input format
+        kwargs (dict): arguments to hub.KerasLayer call
+    '''
+
+    _log_attributes = ('url_or_path', 'features', 'transform_out')
+    _input_type = Stim
+
+    def __init__(self, url_or_path, features=None,
+                 transform_out=None, transform_inp=None,
+                 **kwargs):
+        verify_dependencies(['tensorflow_hub'])
+        self.model = hub.KerasLayer(url_or_path, **kwargs)
+        self.url_or_path = url_or_path
+        self.features = listify(features)
+        self.transform_out = transform_out
+        self.transform_inp = transform_inp
+        super().__init__()
+
+    def get_feature_names(self, out):
```

Comment:
This is currently publicly exposed, so we might want to move the check for `self.features` inside here (instead of doing it in `_extract`) and use that if available. Otherwise a user might naively call `get_feature_names` expecting to get the stored feature names, and instead they'll get the naive enumeration of `feature_*`.

Alternatively, if it's not meant to be public, maybe rename to `_get_feature_names`.

Reply:
thanks for the feedback and suggestion, Tal! Moved check for `self.features` inside `get_feature_names`.